File: day-5b/src/main.py
from __future__ import annotations

import collections
import dataclasses as dc
import enum
from collections.abc import Iterable, Iterator
from functools import cached_property
from itertools import batched, islice
from operator import attrgetter
from typing import Self

get_range_start = attrgetter("start")
get_range_stop = attrgetter("stop")


class GardeningThing(enum.Enum):
    seed = enum.auto()
    soil = enum.auto()
    fertilizer = enum.auto()
    water = enum.auto()
    light = enum.auto()
    temperature = enum.auto()
    humidity = enum.auto()
    location = enum.auto()


@dc.dataclass(kw_only=True, slots=True, frozen=True)
class MapKind:
    source: GardeningThing
    destination: GardeningThing

    def __repr__(self) -> str:
        return f"MapKind({self.source.name}-to-{self.destination.name})"

    @classmethod
    def from_description(cls, description: str) -> Self:
        source, _, destination = description.partition("-to-")
        return cls(
            source=GardeningThing[source], destination=GardeningThing[destination]
        )


@dc.dataclass(frozen=True)
class InputDataRow:
    destination_start: int
    source_start: int
    range_length: int

    @cached_property
    def source_range(self) -> range:
        return range(self.source_start, self.source_start + self.range_length)

    def convert_single(self, item: int) -> int:
        assert (item in self.source_range) or item == self.source_range.stop
        difference = item - self.source_start
        return self.destination_start + difference

    def convert_range(self, r: range) -> range:
        start = self.convert_single(r.start)
        stop = self.convert_single(r.stop)
        return range(start, stop)


@dc.dataclass(slots=True, kw_only=True, frozen=True)
class InputMap:
    kind: MapKind
    rows: tuple[InputDataRow, ...] = dc.field(repr=False)


def find_range_overlap(x: range, y: range) -> range:
    return range(max(x.start, y.start), min(x.stop, y.stop))


@dc.dataclass(slots=True, kw_only=True, frozen=True)
class RangeMap:
    kind: MapKind
    mapping: dict[range, range]


def sliding_pairs_window(iterable: Iterable[range]) -> Iterator[tuple[range, range]]:
    """Adapted from the itertools sliding_window recipe"""
    # sliding_window('ABCDEFG', 4) --> ABCD BCDE CDEF DEFG
    it = iter(iterable)
    window = collections.deque(islice(it, 1), maxlen=2)
    for x in it:
        window.append(x)
        yield tuple(window)  # type: ignore


def _check_range_mapping_consistency(
    initial: dict[range, range], transformed: dict[range, range]
) -> None:
    """Some debug assertions to check correctness"""
    assert (
        min(initial, key=get_range_start).start
        == min(transformed, key=get_range_start).start
    )
    assert (
        max(initial, key=get_range_stop).stop
        == max(transformed, key=get_range_stop).stop
    )
    assert sum(len(key) for key in initial) == sum(len(key) for key in transformed)
    assert len(transformed) >= len(initial)


def progress_range_pair(
    pair: tuple[range, range], input_map: InputMap
) -> dict[range, range]:
    range_mapping: dict[range, range] = {}
    seed_range, intermediate_range = pair
    assert len(seed_range) == len(intermediate_range)
    for row in input_map.rows:
        overlap = find_range_overlap(intermediate_range, row.source_range)
        if overlap:
            new_key = range(
                seed_range.start + (overlap.start - intermediate_range.start),
                seed_range.stop - (intermediate_range.stop - overlap.stop),
            )
            range_mapping[new_key] = row.convert_range(overlap)

    if not range_mapping:
        return {seed_range: intermediate_range}
    keys = sorted(range_mapping, key=get_range_start)
    first_key, last_key = keys[0], keys[-1]
    if seed_range.start < first_key.start:
        startfill = range(seed_range.start, first_key.start)
        startfill_value = range(
            intermediate_range.start, intermediate_range.start + len(startfill)
        )
        range_mapping[startfill] = startfill_value
    if seed_range.stop > last_key.stop:
        endfill = range(last_key.stop, seed_range.stop)
        endfill_value = range(
            intermediate_range.stop - len(endfill), intermediate_range.stop
        )
        range_mapping[endfill] = endfill_value
    for this_range, next_range in sliding_pairs_window(keys):
        if this_range.stop == next_range.start:
            continue
        in_between = range(this_range.stop, next_range.start)
        in_between_value = range(
            intermediate_range.start + (in_between.start - seed_range.start),
            intermediate_range.stop - (seed_range.stop - in_between.stop),
        )
        range_mapping[in_between] = in_between_value
    _check_range_mapping_consistency(dict([pair]), range_mapping)
    if len(range_mapping) > 1:
        assert any(key != value for key, value in range_mapping.items())
    return range_mapping


def progress_range_map(current_range_map: RangeMap, input_data: InputData) -> RangeMap:
    range_mapping: dict[range, range] = {}
    relevant_input_map = next(
        m
        for m in input_data.maps
        if m.kind.source is current_range_map.kind.destination
    )
    for pair in current_range_map.mapping.items():
        range_mapping |= progress_range_pair(pair, relevant_input_map)
    kind = MapKind(
        source=GardeningThing.seed, destination=relevant_input_map.kind.destination
    )
    _check_range_mapping_consistency(current_range_map.mapping, range_mapping)
    return RangeMap(kind=kind, mapping=range_mapping)


@dc.dataclass(slots=True, kw_only=True, frozen=True)
class InputData:
    seed_ranges: tuple[range, ...]
    maps: tuple[InputMap, ...]


def seedrange_to_locationrange(input_data: InputData) -> RangeMap:
    kind = MapKind(source=GardeningThing.seed, destination=GardeningThing.seed)
    initial_range_map = {r: r for r in input_data.seed_ranges}
    range_map = RangeMap(kind=kind, mapping=initial_range_map)

    while range_map.kind.destination is not GardeningThing.location:
        range_map = progress_range_map(range_map, input_data)

    return range_map


def parse_input(filename: str) -> InputData:
    with open(filename) as f:
        puzzle_input = f.read()
    unparsed_seeds, *unparsed_maps = puzzle_input.split("\n\n")
    pairs = batched(map(int, unparsed_seeds.split(" ")[1:]), 2)
    seed_ranges = tuple(range(a, a + b) for a, b in pairs)
    maps: list[InputMap] = []
    for unparsed_map in unparsed_maps:
        first_line, *rest = unparsed_map.splitlines()
        kind = MapKind.from_description(first_line.partition(" ")[0])
        rows = tuple(InputDataRow(*map(int, line.split())) for line in rest)
        maps.append(InputMap(kind=kind, rows=rows))
    return InputData(seed_ranges=seed_ranges, maps=tuple(maps))


def solve(filename: str) -> int:
    input_data = parse_input(filename)
    range_map = seedrange_to_locationrange(input_data)
    lowest_location_range = min(range_map.mapping.values(), key=get_range_start)
    return lowest_location_range.start


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-5b/src/main.rs
use std::cmp::{max, min};
use std::collections::HashMap;
use std::fs::read_to_string;
use std::iter::zip;
use std::ops::Range;
use std::str::FromStr;

use ::anyhow;
use anyhow::{anyhow, bail, Context, Result};

#[derive(PartialEq, Clone, Copy)]
enum GardeningThing {
    Seed,
    Soil,
    Fertilizer,
    Water,
    Light,
    Temperature,
    Humidity,
    Location,
}

impl FromStr for GardeningThing {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> anyhow::Result<Self, Self::Err> {
        match s {
            "seed" => Ok(GardeningThing::Seed),
            "soil" => Ok(GardeningThing::Soil),
            "fertilizer" => Ok(GardeningThing::Fertilizer),
            "water" => Ok(GardeningThing::Water),
            "light" => Ok(GardeningThing::Light),
            "temperature" => Ok(GardeningThing::Temperature),
            "humidity" => Ok(GardeningThing::Humidity),
            "location" => Ok(GardeningThing::Location),
            _ => Err(anyhow!("Unknown gardening thing {}", s)),
        }
    }
}

struct MapKind {
    source: GardeningThing,
    destination: GardeningThing,
}

impl FromStr for MapKind {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.split("-").collect::<Vec<&str>>()[..] {
            [source_description, _, destination_description] => Ok(MapKind {
                source: source_description.parse()?,
                destination: destination_description.parse()?,
            }),
            _ => Err(anyhow!("Can't construct a MapKind from {}", s)),
        }
    }
}

struct InputDataRow {
    destination_start: u64,
    source_start: u64,
    range_length: u64,
}

impl InputDataRow {
    fn source_range(&self) -> Range<u64> {
        self.source_start..(self.source_start + self.range_length)
    }

    fn convert_single(&self, item: u64) -> u64 {
        let source_range = self.source_range();
        assert!(source_range.contains(&item) || item == source_range.end);
        let difference = item - self.source_start;
        self.destination_start + difference
    }

    fn convert_range(&self, r: Range<u64>) -> Range<u64> {
        let start = self.convert_single(r.start);
        let end = self.convert_single(r.end);
        start..end
    }
}

impl FromStr for InputDataRow {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s
            .split_whitespace()
            .map(|s| s.parse())
            .collect::<Result<Vec<u64>, _>>()?[..]
        {
            [destination_start, source_start, range_length] => Ok(InputDataRow {
                destination_start,
                source_start,
                range_length,
            }),
            _ => Err(anyhow!("Couldn't construct an InputDataRow from {}", s)),
        }
    }
}

struct InputMap {
    kind: MapKind,
    rows: Vec<InputDataRow>,
}

impl FromStr for InputMap {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match &s.split("\n").collect::<Vec<&str>>()[..] {
            [first_line, unparsed_rows @ ..] => {
                if unparsed_rows.len() <= 1 {
                    bail!("Expected there to be two or more rows in the map!")
                }
                let kind_description = first_line
                    .split(" ")
                    .next()
                    .context("Expected the first line to have two or more words!")?;
                let kind: MapKind = kind_description.parse()?;
                let rows = unparsed_rows
                    .iter()
                    .map(|s| s.parse())
                    .collect::<Result<Vec<InputDataRow>>>()?;
                Ok(InputMap { kind, rows })
            }
            _ => Err(anyhow!("Couldn't construct an InputMap from {}", s)),
        }
    }
}

fn find_range_overlap(x: &Range<u64>, y: &Range<u64>) -> Range<u64> {
    max(x.start, y.start)..min(x.end, y.end)
}

struct RangeMap {
    kind: MapKind,
    mapping: HashMap<Range<u64>, Range<u64>>,
}

#[cfg(debug_assertions)]
fn _check_range_mapping_consistency(
    initial: &HashMap<Range<u64>, Range<u64>>,
    transformed: &HashMap<Range<u64>, Range<u64>>,
) {
    assert_eq!(
        initial.keys().map(|r| r.start).min().unwrap(),
        transformed.keys().map(|r| r.start).min().unwrap()
    );
    assert_eq!(
        initial.keys().map(|r| r.end).max().unwrap(),
        transformed.keys().map(|r| r.end).max().unwrap()
    );
    assert_eq!(
        initial.keys().map(|r| r.end - r.start).sum::<u64>(),
        transformed.keys().map(|r| r.end - r.start).sum::<u64>()
    );
    assert!(transformed.len() >= initial.len());
}

fn progress_range_pair(
    pair: (&Range<u64>, &Range<u64>),
    input_map: &InputMap,
) -> HashMap<Range<u64>, Range<u64>> {
    let mut range_mapping = HashMap::new();
    let (ref seed_range, ref intermediate_range) = pair;
    debug_assert_eq!(
        (seed_range.end - seed_range.start),
        (intermediate_range.end - intermediate_range.start)
    );
    for row in &input_map.rows {
        let overlap = find_range_overlap(&intermediate_range, &row.source_range());
        if overlap.end > overlap.start {
            let new_key_start = seed_range.start + (overlap.start - intermediate_range.start);
            let new_key_end = seed_range.end - (intermediate_range.end - overlap.end);
            let new_key = new_key_start..new_key_end;
            range_mapping.insert(new_key, row.convert_range(overlap));
        };
    }
    if range_mapping.len() == 0 {
        return HashMap::from_iter([(
            seed_range.to_owned().to_owned(),
            intermediate_range.to_owned().to_owned(),
        )]);
    };
    let mut keys = range_mapping
        .keys()
        .map(|k| k.clone())
        .collect::<Vec<Range<u64>>>();
    keys.sort_unstable_by_key(|r| r.start);
    let (first_key, last_key) = match &keys[..] {
        [first_key, ..] => (first_key, &keys[keys.len() - 1]),
        _ => panic!(),
    };
    if seed_range.start < first_key.start {
        let startfill = seed_range.start..first_key.start;
        let startfill_value =
            intermediate_range.start..(intermediate_range.start + startfill.end - startfill.start);
        range_mapping.insert(startfill, startfill_value);
    }
    if seed_range.end > last_key.end {
        let endfill = last_key.end..seed_range.end;
        let endfill_value =
            (intermediate_range.end + endfill.start - endfill.end)..intermediate_range.end;
        range_mapping.insert(endfill, endfill_value);
    }
    for (this_range, next_range) in zip(&keys[..], &keys[1..]) {
        if this_range.end == next_range.start {
            continue;
        };
        let in_between = this_range.end..next_range.start;
        let in_between_value_start =
            intermediate_range.start + (in_between.start - seed_range.start);
        let in_between_value_end = intermediate_range.end - (seed_range.end - in_between.end);
        range_mapping.insert(in_between, in_between_value_start..in_between_value_end);
    }
    if cfg!(debug_assertions) {
        _check_range_mapping_consistency(
            &HashMap::from_iter([(
                seed_range.to_owned().to_owned(),
                intermediate_range.to_owned().to_owned(),
            )]),
            &range_mapping,
        );
    }
    if range_mapping.len() > 1 {
        debug_assert!(range_mapping.iter().any(|(key, value)| key != value));
    }
    range_mapping
}

fn progress_range_map(current_range_map: RangeMap, input_data: &InputData) -> RangeMap {
    let mut range_mapping = HashMap::<Range<u64>, Range<u64>>::new();
    let relevant_input_map = input_data
        .maps
        .iter()
        .filter(|m| m.kind.source == current_range_map.kind.destination)
        .next()
        .expect("Expected input_data.maps to have length of at least 1!");
    for pair in &current_range_map.mapping {
        for (key, value) in progress_range_pair(pair, relevant_input_map) {
            range_mapping.insert(key, value);
        }
    }
    let kind = MapKind {
        source: GardeningThing::Seed,
        destination: relevant_input_map.kind.destination,
    };
    if cfg!(debug_assertions) {
        _check_range_mapping_consistency(&current_range_map.mapping, &range_mapping);
    }
    RangeMap {
        kind,
        mapping: range_mapping,
    }
}

struct InputData {
    seed_ranges: Vec<Range<u64>>,
    maps: Vec<InputMap>,
}

impl FromStr for InputData {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match &s.replace("\r\n", "\n").split("\n\n").collect::<Vec<&str>>()[..] {
            [unparsed_seeds, unparsed_maps @ ..] => {
                if unparsed_maps.len() <= 1 {
                    bail!("Expected there to be 2 or more maps!")
                }
                let seed_ranges = parse_seed_ranges_from_input(&unparsed_seeds)?;
                let maps = unparsed_maps
                    .iter()
                    .map(|s| s.parse())
                    .collect::<Result<Vec<InputMap>>>()?;
                Ok(InputData { seed_ranges, maps })
            }
            _ => Err(anyhow!("Couldn't parse the input data!")),
        }
    }
}

fn parse_input(filename: &str) -> InputData {
    let puzzle_input =
        read_to_string(filename).expect(format!("Expected file {} to exist", filename).as_str());
    puzzle_input.parse().unwrap()
}

fn seedrange_to_locationrange(input_data: InputData) -> RangeMap {
    let kind = MapKind {
        source: GardeningThing::Seed,
        destination: GardeningThing::Seed,
    };
    let initial_range_map = HashMap::from_iter(
        input_data
            .seed_ranges
            .iter()
            .map(|r| (r.clone(), r.clone())),
    );
    let mut range_map = RangeMap {
        kind,
        mapping: initial_range_map,
    };
    while range_map.kind.destination != GardeningThing::Location {
        range_map = progress_range_map(range_map, &input_data)
    }
    range_map
}

fn parse_seed_ranges_from_input(seed_description: &str) -> Result<Vec<Range<u64>>> {
    Ok(seed_description
        .split(" ")
        .skip(1)
        .map(|s| s.parse::<u64>())
        .collect::<Result<Vec<u64>, _>>()?
        .chunks(2)
        .map(|chunk| chunk[0]..(chunk[0] + chunk[1]))
        .collect())
}

fn solve(filename: &str) -> u64 {
    let input_data = parse_input(filename);
    let range_map = seedrange_to_locationrange(input_data);
    range_map
        .mapping
        .values()
        .min_by_key(|r| r.start)
        .unwrap()
        .start
}

fn main() {
    println!("{}", solve("input.txt"));
}
